subprojects/version-control/src/main/java/org/gradle/vcs/git/GitVersionControlSystem.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.vcs.git;
+
+import org.eclipse.jgit.api.CloneCommand;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.vcs.VersionControlSpec;
+import org.gradle.vcs.VersionControlSystem;
+
+import java.io.File;
+
+/**
+ * A Git {@link VersionControlSystem} implementation.
+ */
+public class GitVersionControlSystem implements VersionControlSystem {
```

Comment:
We should think about what kind of guarantees we want to provide/require with this.  I think it would make sense to provide our own locking around these implementations instead of requiring a VCS implementation to provide its own.

We should also provide build operations around this and not force VCS implementations to implement that.

We may also want to have our own caching fit in here somehow (e.g., put a copy of a bare repository into ~/.gradle), but I haven't thought about how that would work yet. 

As an initial cut, I think it's fine to embed the locking inside here or have it as a wrapper around it.  Either way as long as we have it.

Reply:
I think I've addressed everything you mentioned in this review **except** for this comment. I would really like to discuss the following topics with you before I add them to the branch:
    * Locking
        * Especially at what granularity we should implement locking.
    * Build Operations
        * I'm just looking for some pointers as I haven't implemented this before.
    * Caching
        * I think this would be better suited for a follow-on change. I have some ideas for what to do for Git repositories, but I think we can get things working and then sweep back through and make them efficient.